File: PaintersDilemma.py
def parser():
    dataInput = list(input().split(' '))

    integer_map = map(int, dataInput)

    dataInputInts = list(integer_map)
    return dataInputInts


def switchBrushPaints(i_Scenario_Sequence, i_Current_Index):
    global minimal_Amount
    global first_Brush
    global second_Brush

    movingIndex = i_Current_Index
    foundAnyBrush = False

    for movingIndex in range(movingIndex, len(i_Scenario_Sequence)):
        if i_Scenario_Sequence[movingIndex] == first_Brush:
            second_Brush = i_Scenario_Sequence[i_Current_Index]
            minimal_Amount += 1
            foundAnyBrush = True
            break
        if i_Scenario_Sequence[movingIndex] == second_Brush:
            first_Brush = i_Scenario_Sequence[i_Current_Index]
            minimal_Amount += 1
            foundAnyBrush = True
            break
    if not foundAnyBrush:
        first_Brush = i_Scenario_Sequence[i_Current_Index]
        minimal_Amount += 1


def computeMinimalAmount(i_Scenario):
    global minimal_Amount
    global first_Brush
    global second_Brush

    minimal_Amount = 1
    first_Brush = i_Scenario[0]
    second_Brush = 0
    i = 0
    scenario_Size = len(i_Scenario)

    for i in range(1, scenario_Size):
        if(i_Scenario[i] != first_Brush):
            second_Brush = i_Scenario[i]
            minimal_Amount += 1
            break
        i += 1

    i += 1
    for i in range(i, scenario_Size):
        if i_Scenario[i] == first_Brush or i_Scenario[i] == second_Brush:
            continue
        else:
            switchBrushPaints(i_Scenario, i)

    return minimal_Amount


amount_Of_Scenarios = int(input())

listOfColorSequences = []
for x in range(0, amount_Of_Scenarios):
    input()
    colorSequence = []
    colorSequence = parser()
    listOfColorSequences.append(colorSequence)

for scenario in listOfColorSequences:
    answer = computeMinimalAmount(scenario)
    print(answer)

File: PaintersDilemma.cs
using System;
using System.IO;
using System.Collections.Generic;

namespace Painters_Dilemma
{
    class Solution
    {
        public static List<int> parseScenarioInput()
        {
            List<int> colorSequence = new List<int>();
            int color = 0;
            while (true)
            {
                int currentChar = Console.Read();
                if (currentChar == 32 || currentChar == 13)
                {
                    colorSequence.Add(color);
                    color = 0;
                    continue;
                }
                if (currentChar == 10)
                {
                    break;
                }
                color = (color * 10) + (currentChar - 48);
            }
            return colorSequence;
        }

        public static void switchBrushPaints(ref int io_MinimalAmount, List<int> i_ColorSequence, ref int io_FirstBrush, ref int io_SecondBrush, int i_CurrentIndex)
        {
            int movingIndex = i_CurrentIndex;
            bool foundAnyBrush = false;
            for (; movingIndex < i_ColorSequence.Count; movingIndex++)
            {
                if (i_ColorSequence[movingIndex] == io_FirstBrush)
                {
                    io_SecondBrush = i_ColorSequence[i_CurrentIndex];
                    io_MinimalAmount++;
                    foundAnyBrush = true;
                    break;
                }
                if (i_ColorSequence[movingIndex] == io_SecondBrush)
                {
                    io_FirstBrush = i_ColorSequence[i_CurrentIndex];
                    io_MinimalAmount++;
                    foundAnyBrush = true;
                    break;
                }
            }
            if (!foundAnyBrush)
            {
                io_FirstBrush = i_ColorSequence[i_CurrentIndex];
                io_MinimalAmount++;
            }
        }

        public static int computeMinimalAmount(List<int> i_ColorSequence, int i_AmountOfPaintActions)
        {
            int minimalAmount = 1;
            int firstBrush = i_ColorSequence[0];
            int secondBrush = 0;
            int i = 1;

            //deciding secondBrush's paint
            for (; i < i_AmountOfPaintActions; i++)
            {
                if (i_ColorSequence[i] != firstBrush)
                {
                    secondBrush = i_ColorSequence[i];
                    minimalAmount++;
                    break;
                }
            }

            i++;
            //checking for every other possible paint brush switch
            for (; i < i_AmountOfPaintActions; i++)
            {
                if (i_ColorSequence[i] == firstBrush || i_ColorSequence[i] == secondBrush)
                {
                    continue;
                }
                else
                {
                    switchBrushPaints(ref minimalAmount, i_ColorSequence, ref firstBrush, ref secondBrush, i);
                }
            }

            return minimalAmount;
        }

        public static void Main(String[] args)
        {
            Console.SetIn(new StreamReader(Console.OpenStandardInput(8192), Console.InputEncoding, false, 8192));
            int amountOfScenarios = Convert.ToInt32(Console.ReadLine());
            List<List<int>> ListOfColorSequences = new List<List<int>>();

            for (int i = 0; i < amountOfScenarios; i++)
            {
                Console.ReadLine();
                List<int> colorSequence = parseScenarioInput();
                ListOfColorSequences.Add(colorSequence);
            }

            for (int i = 0; i < amountOfScenarios; i++)
            {
                int minimalSwitchesAmount = computeMinimalAmount(ListOfColorSequences[i], ListOfColorSequences[i].Count);
                Console.WriteLine(minimalSwitchesAmount);
            }
            //Console.Read();
        }
    }
}